File: cpp/common/logging.cpp
#include "cpp/common/logging.hpp"

#include <algorithm>
#include <array>
#include <cctype>
#include <cstdint>
#include <fstream>
#include <optional>
#include <stdexcept>
#include <string>
#include <string_view>
#include <utility>
#include <variant>

#include <nlohmann/json.hpp>

#include "cpp/common/clock.hpp"

namespace aegis::common {
namespace {

using nlohmann::json;

/// Substrings that mark a field as holding a credential. Matched against the
/// lowercased key, so `API_KEY`, `db.password` and `awsSecret` all hit.
constexpr std::string_view kRedacted = "[redacted]";

constexpr std::array<std::string_view, 9> kSecretMarkers = {
    "password",   "passwd",  "secret",        "token", "key",
    "credential", "private", "authorization", "auth"};

std::string lowercase(std::string_view text) {
  std::string lowered{text};
  std::ranges::transform(lowered, lowered.begin(),
                         [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return lowered;
}

json to_json_value(const FieldValue& value) {
  return std::visit(
      [](const auto& held) -> json {
        using Held = std::decay_t<decltype(held)>;
        if constexpr (std::is_same_v<Held, std::monostate>) {
          return nullptr;
        } else {
          return held;
        }
      },
      value);
}

}  // namespace

std::string_view to_string(LogLevel level) {
  switch (level) {
    case LogLevel::kTrace:
      return "trace";
    case LogLevel::kDebug:
      return "debug";
    case LogLevel::kInfo:
      return "info";
    case LogLevel::kWarn:
      return "warn";
    case LogLevel::kError:
      return "error";
  }
  return "info";
}

std::optional<LogLevel> parse_log_level(std::string_view name) {
  const auto lowered = lowercase(name);
  if (lowered == "trace") return LogLevel::kTrace;
  if (lowered == "debug") return LogLevel::kDebug;
  if (lowered == "info") return LogLevel::kInfo;
  if (lowered == "warn") return LogLevel::kWarn;
  if (lowered == "error") return LogLevel::kError;
  return std::nullopt;
}

bool is_secret_field(std::string_view name) {
  const auto lowered = lowercase(name);
  return std::ranges::any_of(kSecretMarkers, [&lowered](std::string_view marker) {
    return lowered.find(marker) != std::string::npos;
  });
}

void FileLogSink::write_line(const std::string& line) {
  std::ofstream output{path_, std::ios::app};
  if (!output) {
    throw std::runtime_error("cannot open log sink for append: " + path_);
  }
  output << line << '\n';
}

StructuredLogger::StructuredLogger(std::string name, std::string experiment_id,
                                   const WallClock& clock, LogSink& sink, LogLevel level,
                                   std::string correlation_id)
    : name_(std::move(name)),
      experiment_id_(std::move(experiment_id)),
      clock_(&clock),
      sink_(&sink),
      level_(level),
      correlation_id_(std::move(correlation_id)) {
  if (name_.empty()) {
    throw std::invalid_argument("logger name is required; records must say what emitted them");
  }
  if (experiment_id_.empty()) {
    throw std::invalid_argument(
        "experiment_id is required; a record that cannot be joined to its run is not evidence "
        "of anything");
  }
}

StructuredLogger StructuredLogger::bind(std::string name, std::string correlation_id) const {
  return StructuredLogger{std::move(name), experiment_id_, *clock_,
                          *sink_,          level_,         std::move(correlation_id)};
}

void StructuredLogger::log(LogLevel level, std::string_view message, const Fields& fields) {
  if (static_cast<std::uint8_t>(level) < static_cast<std::uint8_t>(level_)) {
    return;
  }

  // nlohmann's ordered_json preserves insertion order, so the key order below is
  // the key order on disk. Two identical runs therefore produce identical bytes,
  // which is what makes a log file usable as determinism-harness input.
  auto record = nlohmann::ordered_json::object();
  record.emplace("schema_version", kLogSchemaVersion);
  record.emplace("timestamp_ns", clock_->now_utc());
  record.emplace("level", std::string{to_string(level)});
  record.emplace("logger", name_);
  record.emplace("message", std::string{message});
  record.emplace("experiment_id", experiment_id_);
  record.emplace("sequence", sequence_);
  if (!correlation_id_.empty()) {
    record.emplace("correlation_id", correlation_id_);
  }
  if (!fields.empty()) {
    auto payload = nlohmann::ordered_json::object();
    for (const auto& [key, value] : fields) {  // std::map iterates in key order
      payload.emplace(key, is_secret_field(key) ? json(kRedacted) : to_json_value(value));
    }
    record.emplace("fields", std::move(payload));
  }

  ++sequence_;
  sink_->write_line(record.dump());
}

}  // namespace aegis::common

File: python/common/logging.py
"""Structured JSON Lines logging (AEGIS-232).

Four properties, each answering a specific way logs stop being useful:

**Machine-readable.** One JSON object per line, validated against
``configs/schemas/log_record.v1.json``. Free-text logs cannot be joined,
filtered or diffed, so they cannot serve as evidence.

**Correlated.** Every record carries ``experiment_id`` — the same field the
message envelope carries — plus an optional ``correlation_id`` for one causal
chain. Without them a log line and the event that produced it can only be
matched by timestamp and hope.

**Deterministic.** The clock is injected and a per-logger sequence number breaks
timestamp ties, so a fixture run twice produces byte-identical output. That is
what lets a log file be an input to the determinism harness rather than a source
of spurious diffs.

**Secret-free.** Field values are redacted by key name before serialization.
A logger is the most common way a credential reaches disk, and the check has to
live where the record is built rather than in a review convention.

The logger is an instance. There is no module-level default and no
``get_logger()`` singleton: a process-global logger is shared mutable state
reachable from every book partition, which is what AEGIS-047 exists to prevent.
"""

from __future__ import annotations

import json
import re
from collections.abc import Mapping
from dataclasses import dataclass
from enum import IntEnum
from pathlib import Path
from typing import Any, Protocol, TextIO

from common.clock import Nanos, WallClock

SCHEMA_VERSION = 1
SCHEMA_PATH = "configs/schemas/log_record.v1.json"

REDACTED = "[redacted]"

# Field names whose values never reach a log record. Matched case-insensitively
# against the whole key, so `api_key`, `AWS_SECRET` and `db.password` all hit.
SECRET_KEY_PATTERN = re.compile(
    r"(?i)(password|passwd|secret|token|api[_-]?key|access[_-]?key|credential|private[_-]?key|authorization)"
)


class Level(IntEnum):
    """Ordered so a threshold comparison is a single integer test."""

    TRACE = 10
    DEBUG = 20
    INFO = 30
    WARN = 40
    ERROR = 50

    @property
    def label(self) -> str:
        return self.name.lower()

    @classmethod
    def parse(cls, name: str) -> Level:
        try:
            return cls[name.strip().upper()]
        except KeyError:
            valid = ", ".join(level.label for level in cls)
            raise ValueError(f"unknown log level {name!r}; expected one of {valid}") from None


class Sink(Protocol):
    """Where records go. A sink writes one line and does not interpret it."""

    def write_line(self, line: str) -> None: ...


class StreamSink:
    """Write records to an open text stream."""

    def __init__(self, stream: TextIO) -> None:
        self._stream = stream

    def write_line(self, line: str) -> None:
        self._stream.write(line + "\n")
        self._stream.flush()


class ListSink:
    """Collect records in memory. The sink tests and fixtures use."""

    def __init__(self) -> None:
        self.lines: list[str] = []

    def write_line(self, line: str) -> None:
        self.lines.append(line)

    @property
    def records(self) -> list[dict[str, Any]]:
        return [json.loads(line) for line in self.lines]


class FileSink:
    """Append records to a file, creating parent directories as needed."""

    def __init__(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        self._path = path

    def write_line(self, line: str) -> None:
        with self._path.open("a", encoding="utf-8") as handle:
            handle.write(line + "\n")


def redact(fields: Mapping[str, Any]) -> dict[str, Any]:
    """Replace secret-looking values, keeping the key so the shape stays readable."""
    return {
        key: (REDACTED if SECRET_KEY_PATTERN.search(key) else value) for key, value in fields.items()
    }


def _coerce_value(value: Any) -> str | int | float | bool | None:
    """Force a field value into a scalar the schema permits.

    Anything structured is rendered to its repr rather than nested. Unbounded
    nested payloads on a hot path are how a log becomes the slowest part of a
    system, and a record whose shape varies per call cannot be queried.
    """
    if value is None or isinstance(value, str | int | float | bool):
        return value
    return repr(value)


@dataclass(frozen=True)
class LogRecord:
    """One structured record, in the field order the schema declares."""

    schema_version: int
    timestamp_ns: Nanos
    level: str
    logger: str
    message: str
    experiment_id: str
    sequence: int
    correlation_id: str | None = None
    fields: Mapping[str, Any] | None = None

    def to_json(self) -> str:
        """Serialize with a fixed key order, so two identical runs produce identical bytes."""
        payload: dict[str, Any] = {
            "schema_version": self.schema_version,
            "timestamp_ns": self.timestamp_ns,
            "level": self.level,
            "logger": self.logger,
            "message": self.message,
            "experiment_id": self.experiment_id,
            "sequence": self.sequence,
        }
        if self.correlation_id is not None:
            payload["correlation_id"] = self.correlation_id
        if self.fields:
            payload["fields"] = dict(sorted(self.fields.items()))
        return json.dumps(payload, separators=(",", ":"), ensure_ascii=False)


class StructuredLogger:
    """A logger instance bound to one experiment, clock and sink."""

    def __init__(
        self,
        name: str,
        experiment_id: str,
        clock: WallClock,
        sink: Sink,
        level: Level = Level.INFO,
        correlation_id: str | None = None,
    ) -> None:
        if not name:
            raise ValueError("logger name is required; records must say what emitted them")
        if not experiment_id:
            raise ValueError(
                "experiment_id is required; a record that cannot be joined to its run "
                "is not evidence of anything"
            )
        self._name = name
        self._experiment_id = experiment_id
        self._clock = clock
        self._sink = sink
        self._level = level
        self._correlation_id = correlation_id
        self._sequence = 0

    @property
    def sequence(self) -> int:
        return self._sequence

    @property
    def level(self) -> Level:
        return self._level

    def bind(self, *, name: str | None = None, correlation_id: str | None = None) -> StructuredLogger:
        """Derive a child logger sharing this logger's clock, sink and experiment.

        The child keeps its own sequence counter: sequence numbers order the
        records of one emitter, and sharing a counter across components would
        make the ordering depend on interleaving.
        """
        return StructuredLogger(
            name=name or self._name,
            experiment_id=self._experiment_id,
            clock=self._clock,
            sink=self._sink,
            level=self._level,
            correlation_id=correlation_id if correlation_id is not None else self._correlation_id,
        )

    def log(self, level: Level, message: str, **fields: Any) -> LogRecord | None:
        if level < self._level:
            return None
        record = LogRecord(
            schema_version=SCHEMA_VERSION,
            timestamp_ns=self._clock.now_utc(),
            level=level.label,
            logger=self._name,
            message=message,
            experiment_id=self._experiment_id,
            sequence=self._sequence,
            correlation_id=self._correlation_id,
            fields={key: _coerce_value(value) for key, value in redact(fields).items()} or None,
        )
        self._sequence += 1
        self._sink.write_line(record.to_json())
        return record

    def trace(self, message: str, **fields: Any) -> LogRecord | None:
        return self.log(Level.TRACE, message, **fields)

    def debug(self, message: str, **fields: Any) -> LogRecord | None:
        return self.log(Level.DEBUG, message, **fields)

    def info(self, message: str, **fields: Any) -> LogRecord | None:
        return self.log(Level.INFO, message, **fields)

    def warn(self, message: str, **fields: Any) -> LogRecord | None:
        return self.log(Level.WARN, message, **fields)

    def error(self, message: str, **fields: Any) -> LogRecord | None:
        return self.log(Level.ERROR, message, **fields)


def load_schema(root: Path) -> dict[str, Any]:
    schema: dict[str, Any] = json.loads((root / SCHEMA_PATH).read_text(encoding="utf-8"))
    return schema
